src/linking/linkModule.ml

```diff
@@ -743,75 +766,206 @@ let mk_i32_global (i : int32) =
-      let name = import.it.item_name in
-      let fun_idx =
-        match find_fun_export name m.exports with
-        | None -> raise (LinkError (Format.sprintf "Can't find export for GOT.func import %s" (Lib.Utf8.encode name)))
-        | Some export_idx -> export_idx.it
-      in
-      let global_idx =
-        if is_global_import import.it.idesc.it then
-          global_idx
-        else
-          raise (LinkError "GOT.func import is not global")
-      in
-      ( Int32.add global_idx (Int32.of_int 1), (global_idx, fun_idx) :: imports )
+      let got_func = resolve_got_func global_index import m in
+      (false, next_index, got_func :: imports)
+    else if import.it.module_name = got_mem_name then
+      let got_mem = resolve_got_mem global_index import m in
+      (false, next_index, got_mem :: imports)
     else
-      let global_idx =
-        if is_global_import import.it.idesc.it then
-          Int32.add global_idx (Int32.of_int 1)
-        else
-          global_idx
+      (* Implementation restriction: No normal imported globals after GOT globals.
+      If this would be required in the future, the GOT globals cannot simply be
+      moved to the beginning of the module's global section because the global
+      indices would then change and the global accesses in the AST would need
+      to be patched. *)
+      let continue_index =
+        if is_global_import import.it.idesc.it then 
+          (assert allow_normal_globals;    
+          next_index)
+        else global_index
       in
-      ( global_idx, imports )
+      (allow_normal_globals, continue_index, imports)
   in
-
-  (* (global index, function index) list *)
-  let (_, got_func_imports) =
-    List.fold_left get_got_func_import (0l, []) m.imports
+  let (_, _, got_imports) =
+    List.fold_left get_got_import (true, 0l, []) m.imports
   in
+  got_imports
 
-  got_func_imports
+(* For each GOT.mem access, we compute the absolute address of the data pointer.
+   This is done by adding the library memory base (`lib_heap_start`) to the offset
+   that is stored in the exported global that corresponds to the GOT.mem import. *)
+let patch_got_mem_accesses got_mem_imports memory_base = fun m ->
+  let phrase_one_to_many f x = List.map (fun y -> { x with it = y }) (f x.it) in
 
-(* Add functions imported from GOT.func to the table, replace GOT.func imports
-   with globals to the table indices.
+  let find_got_mem global_index =
+    List.find_opt (fun (index, _, _) -> index = global_index) got_mem_imports
+  in
 
-   `tbe_size` is the size of the table in the merged module before adding
-   GOT.func functions. *)
-let replace_got_func_imports (tbl_size : int32) (imports : (int32 * int32) list) (m : module_') : module_' =
-  (* null check to avoid adding empty elem section *)
-  if imports = [] then
-    m
+  (* Computes the absolute address of the GOT.mem data pointer *)
+  let data_pointer exported_global_index = if uses_memory64 m then
```

Comment:
Not that you should do this now, but I wonder if it would be possible to just bump the (mutable?) globals by the offset using an initialization function, rather than recomputing the absolute address at each read. Another advantage, apart from perf, would be that any debug info would remain valid (rather than disturbing the instruction offset in the code.

Reply:
A good idea. If okay, I'll remember for later exploration, as GOT.mem are relatively seldom.